File: sort/insertion.c
#include <stdio.h>
#include <stdlib.h>

/********* DEFINED CONSTANTS *********/
#define   MAX          50

/********* FUNCTION DECLARATION *********/
void insertion_ascend(int n, int array[]);
void insertion_decend(int n, int array[]);
void ciltable(int end, int start, int array[]);

/********* MAIN STARTS HERE ********/
int main()
{
   int        i, n, choice;
   int        array[MAX];

   printf("Enter the no. of elements in table: ");
   scanf("%d", &n);

   printf("Enter the elements in table: \n");
   for (i = 0; i < n; i++)
   {
      printf("Enter the integer: ");
      scanf("%d", &array[i]);
   }

   printf("Select an option: \n");
   printf("1) In ascending order.\n");
   printf("2) In dedcending order.\n");
   scanf("%d", &choice);

   if (choice == 1)
   {
      insertion_ascend(n, array);
   }
   else if (choice == 2)
   {
      insertion_decend(n, array);
   }
   else
   {
      printf("INVALID OPTION.\n");
      exit(1);
   }

   printf("Sorted list: \n[");
   for (i = 0; i < n; i++)
   {
      printf("%d, ", array[i]);
   }
   printf("]\n");

   exit(0);
}

/********* FUNCTION DEFINITION *********/
void insertion_ascend(int n, int array[])
{
   int        i, j, ind;

   for (i = 0; i < n; i++)
   {
      ind = i+1;
      for (j = 0; j < i+1; j++)
      {
          if (array[ind] < array[j])
          {
             ciltable(ind, j, array);
          }
      }
   }

   return ;
}

void insertion_decend(int n, int array[])
{
   int        i, j, ind;

   for (i = 0; i < n; i++)
   {
      ind = i+1;
      for (j = 0; j < i+1; j++)
      {
          if (array[ind] > array[j])
          {
             ciltable(ind, j, array);
          }
      }
   }

   return ;
}

void ciltable(int end, int start, int array[])
{
   int        i, key;

   key = array[end];
   for (i = end; i > start; i--)
   {
      array[i] = array[i-1];
   }
   array[start] = key;

   return ;
}

File: sort/insertion.py
#!/usr/bin/python3.6

def ciltable(start, end, array):
    key = array[start]

    for i in range(start, end, -1):
        array[i] = array[i-1]
    array[end] = key
    return

def insertion_ascend(n, array):
    for i in range(n):
        ind = i
        for j in range(i+1):
            if array[ind] < array[j]:
                ciltable(ind, j, array)

    return ;

def insertion_descend(n, array):
    for i in range(n):
        ind = i
        for j in range(i+1):
            if array[ind] > array[j]:
                ciltable(ind, j, array)

    return ;

n = int(input("Enter the total no. of elements: "))
array = list()

for i in range(n):
    array.append(int(input("Enter the integer: ")))

opt = int(input("Select an option:\n1) Ascending order\n2) Descending order\n"))

if opt == 1:
    insertion_ascend(n, array)
else:
    insertion_descend(n, array)
print("Sorted list: \n", array)
